src/library/export/dlglibraryexport.cpp

```diff
@@ -0,0 +1,202 @@
+#include "library/export/dlglibraryexport.h"
+
+#include <QFileDialog>
+#include <QFormLayout>
+#include <QGridLayout>
+#include <QHBoxLayout>
+#include <QLabel>
+#include <QPushButton>
+#include <QStandardPaths>
+
+#include <djinterop/enginelibrary.hpp>
+
+#include "library/crate/crateid.h"
+#include "library/export/engineprimeexportrequest.h"
+#include "library/trackcollection.h"
+#include "library/trackcollectionmanager.h"
+
+static const QString DefaultMixxxExportDirName = "mixxx-export";
+
+namespace el = djinterop::enginelibrary;
+
+namespace mixxx {
+
+DlgLibraryExport::DlgLibraryExport(
+        QWidget* parent, UserSettingsPointer pConfig, TrackCollectionManager& trackCollectionManager)
+        : QDialog(parent), m_pConfig{pConfig}, m_trackCollectionManager{trackCollectionManager} {
+    // Selectable list of crates from the Mixxx library.
+    m_pCratesList = make_parented<QListWidget>();
+    m_pCratesList->setSelectionMode(QListWidget::ExtendedSelection);
+    {
+        // Populate list of crates.
+        auto crates = m_trackCollectionManager.internalCollection()->crates().selectCrates();
+        Crate crate;
+        while (crates.populateNext(&crate)) {
+            auto pItem = std::make_unique<QListWidgetItem>(crate.getName());
+            pItem->setData(Qt::UserRole, crate.getId().value());
+            m_pCratesList->addItem(pItem.release());
+        }
+    }
+
+    // Read-only text fields showing key directories for export.
+    m_pBaseDirectoryTextField = make_parented<QLineEdit>();
+    m_pBaseDirectoryTextField->setReadOnly(true);
+    m_pDatabaseDirectoryTextField = make_parented<QLineEdit>();
+    m_pDatabaseDirectoryTextField->setReadOnly(true);
+    m_pMusicDirectoryTextField = make_parented<QLineEdit>();
+    m_pMusicDirectoryTextField->setReadOnly(true);
+
+    // Radio buttons to allow choice between exporting the whole music library
+    // or just tracks in a selection of crates.
+    m_pWholeLibraryRadio = make_parented<QRadioButton>(tr("Entire music library"));
+    m_pWholeLibraryRadio->setChecked(true);
+    this->exportWholeLibrarySelected();
+    connect(m_pWholeLibraryRadio,
+            &QRadioButton::clicked,
+            this,
+            &DlgLibraryExport::exportWholeLibrarySelected);
+    m_pCratesRadio = make_parented<QRadioButton>(tr("Selected crates"));
+    connect(m_pCratesRadio,
+            &QRadioButton::clicked,
+            this,
+            &DlgLibraryExport::exportSelectedCratedSelected);
+
+    // Button to allow ability to browse for the export directory.
+    auto pExportDirBrowseButton = make_parented<QPushButton>(tr("Browse"));
```

Comment:
Same here, parented_ptr but no parent. Don't all these allocations end up as memory leaks?

Reply:
The `pExportDirBrowseButton` is added to a layout later in the function.  In fact, all widgets created via `make_parented<>()` are added to a layout somewhere, and hence all have a parent before the ctor completes.

FWIW, turning on fatal debug assertions does not throw anything with the code as it currently is.

Reply:
It must have a parent immediately or otherwise that's not exception safe. I suspect you can write

```suggestion
    auto pExportDirBrowseButton = make_parented<QPushButton>(tr("Browse"), this);
```